File: 이취코/jeongseon/week1/ch03/큰수의법칙.cpp
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

//M : 더하는 횟수
//k : 같은 인덱스 숫자 연속해서 더하는 한계
//인덱스가 다른데 같은 숫자는 서로 다른 것으로 간주

//1억 = 1초

int n, m, k, sum;
vector<int> v;

int main(){

    cin >> n >> m >> k;

    for(int i=0; i<n; i++){
        int num;
        cin >> num;
        v.push_back(num);
    }

    //정렬하기
    sort(v.begin(), v.end(), greater<>());

    int tmpCnt=0, totalCnt=0;
    while(true){
        if(totalCnt == m ) break;
        if(tmpCnt == k){
            tmpCnt=0;   //초기화
            totalCnt++;
            sum += v[1];
        }else{
            sum += v[0];
            tmpCnt++;
            totalCnt++;
        }
    }

    cout << sum << "\n";

    return 0;
}

File: 이취코/sojeong/week1/ch03/큰수의법칙.py
# 큰 수의 법칙

# 주어진 수들을 M번 더하여 가장 큰 수를 만든다
# 배열의 특정한 인덱스(번호)에 해당하는 수가 연속해서 K번을 초과하여 더해질 수 없다.

# 입력
N, M, K = map(int, input().split())
num = list(map(int, input().split()))

# 내림차순
num.sort(reverse=True)

sum = 0
# 첫번째꺼 K번 더한다
# 두번째꺼 1번 더한다
# 첫번째꺼 K번 더한다
# ...

while M > 0:
    sum += (num[0]*K + num[1]*(int(M!=0)))
    M -= (K + int(M!=0))

print(sum)